dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -0,0 +1,486 @@
+    List<String> pathComponents = mProto.getPathComponentsList();
+    return new UfsUrl(UfsUrlMessage.newBuilder()
+        .setScheme(mProto.getScheme())
+        .setAuthority(mProto.getAuthority())
+        .addAllPathComponents(pathComponents).addPathComponents(childName).build());
+  }
+
+  /**
+   * Returns the full path by connecting the pathComponents list.
+   *
+   * @return a full path string
+   */
+  public String getFullPath() {
+    StringBuilder sb = new StringBuilder();
+    sb.append(PATH_SEPARATOR);
+    // Then sb is not empty.
+    for (int i = 1; i < mProto.getPathComponentsList().size(); i++) {
+      sb.append(mProto.getPathComponentsList().get(i));
+      if (!mProto.getPathComponentsList().get(i).isEmpty()) {
+        sb.append(PATH_SEPARATOR);
+      }
+    }
+    return FilenameUtils.normalizeNoEndSeparator(sb.toString());
+  }
+
+  /**
+   * Translates current UfsUrl object to an AlluxioURI object.
+   *
+   * @return a corresponding AlluxioURI object
+   */
+  public AlluxioURI toAlluxioURI() {
+    return new AlluxioURI(mProto.getScheme(),
+        Authority.fromString(mProto.getAuthority()), getFullPath());
+  }
+
+  /**
+   * Returns the number of elements of the path component of the {@link UfsUrl}.
+   *
+   * @return the depth
+   */
+  public int getDepth() {
+    int pathComponentsSize = getPathComponents().size();
+    Preconditions.checkArgument(pathComponentsSize > 0);
+    // "/" is represented as an empty String "".
+    return getPathComponents().get(0).isEmpty() ? pathComponentsSize - 1 : pathComponentsSize;
+  }
+
+  /**
+   * Gets the final component of the {@link UfsUrl}.
+   *
+   * @return the final component of the {@link UfsUrl}
+   */
+  public String getName() {
+    List<String> pathComponents = getPathComponents();
+    Preconditions.checkArgument(!pathComponents.isEmpty());
+    return pathComponents.get(pathComponents.size() - 1);
+  }
+
+  /**
+   * Returns true if the current UfsUrl is an ancestor of another UfsUrl.
+   * otherwise, return false.
+   * @param ufsUrl potential children to check
+   * @return true the current ufsUrl is an ancestor of the ufsUrl
+   */
+  public boolean isAncestorOf(UfsUrl ufsUrl) throws InvalidPathException {
+    if (!Objects.equals(getAuthority(), ufsUrl.getAuthority())) {
+      return false;
+    }
+    if (!Objects.equals(getScheme(), ufsUrl.getScheme())) {
+      return false;
+    }
+    // TODO(Tony Sun): optimize the performance later
+    // Both of the ufsUrls has the same scheme and authority, so just need to compare their paths.
+    return PathUtils.hasPrefix(PathUtils.normalizePath(ufsUrl.getFullPath(), PATH_SEPARATOR),
+        PathUtils.normalizePath(getFullPath(), PATH_SEPARATOR));
+  }
+
+  /**
+   * Appends additional path elements to the end of an {@link UfsUrl}.
+   *
+   * @param suffix the suffix to add
+   * @return the new {@link UfsUrl}
+   */
+  public UfsUrl join(String suffix) {
```

Comment:
IMO this should be implemented as two separate methods:
1. `join(String singlePathSegment)` which rejects the input if it contains a `/`
2. `join(PathSegments segments)` which does not need to do parsing at all, as it is now the caller's responsibility to parse a path string to a `PathSegments` object.